### PR TITLE
runtime: tablewide profiling api

### PR DESCRIPTION
**Pull request by egor-ryashin**

### CODE REVIEW

**Review by begelundmuller — CHANGES_REQUESTED** [review]

File: runtime/server/table_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	if err != nil {
+		logger.Error("did not connect", zap.Error(err))
+	}
+	defer conn.Close()
+	client := api.NewRuntimeServiceClient(conn)
+	cr, _ := client.Cardinality(context.Background(), &api.CardinalityRequest{
+		InstanceId: "test",
+		TableName:  "test",
+	})
```
[the remark this answers is not in the document]
Comment (egor-ryashin) [reply runtime/server/table_test.go:80]: Yes, I thought about that, testing http right away saves time on writing the same tests for e2e, also better test early than later.

[the remark this answers is not in the document]
Comment (egor-ryashin) [reply runtime/server/table_test.go]: I reduced to server side only, let's think about http testing in the next step

File: runtime/server/server.go [hunk 2]
```diff
@@ -99,10 +99,39 @@ func (s *Server) Serve(ctx context.Context) error {
...
+	rows, err := s.os.Execute(ctx, &drivers.Statement{
+		Query: "select count(*) from " + req.TableName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var count int64
+	for rows.Next() {
+		// note that city can be NULL, so we use the NullString type
```
[the remark this answers is not in the document]
Comment (egor-ryashin) [reply runtime/server/server.go]: removed

File: runtime/server/server.go [hunk 3]
```diff
@@ -99,10 +99,39 @@ func (s *Server) Serve(ctx context.Context) error {
...
+func (s *Server) Serve(ctx context.Context) error {
+	group, _ := s.ServeNoWait(ctx)
 	return group.Wait()
 }
 
+func (s *Server) Cardinality(ctx context.Context, req *api.CardinalityRequest) (*api.CardinalityResponse, error) {
+	bb := s.os == nil
+	s.logger.Info("nil " + fmt.Sprintf("%v", bb))
+
+	rows, err := s.os.Execute(ctx, &drivers.Statement{
```
[the remark this answers is not in the document]
Comment (egor-ryashin) [reply runtime/server/server.go]: changed

File: runtime/server/instances.go [hunk 4]
```diff
@@ -50,7 +50,8 @@ func (s *Server) CreateInstance(ctx context.Context, req *api.CreateInstanceRequ
 	if err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-	_, ok := conn.OLAPStore()
+	os, ok := conn.OLAPStore() // todo create a pool of OLAP stores, add or get from the pool instead
+	s.os = os
```
[the remark this answers is not in the document]
Comment (egor-ryashin) [reply runtime/server/instances.go:54]: changed

File: runtime/api/runtime.proto [hunk 5]
```diff
@@ -228,6 +247,41 @@ service RuntimeService {
...
+}
+
+message ProfileColumnsResponse {
+  repeated ProfileColumn profile_column = 1;
+}
+
+message ProfileColumn {
+  string name = 1;
+  string type = 2;
+  int32 largestStringLength = 3;
```
Comment (begelundmuller) [inline runtime/api/runtime.proto:271]: Should be snake case

File: runtime/server/table_api_test.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	rows.Close()
+	return val
+}
+
+func TestServer_Database(t *testing.T) {
+	server, instanceId, err := getTestServer()
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (begelundmuller) [inline runtime/server/table_api_test.go:29]: Can use `require.NoError(t, err)` for these kinds of checks

File: runtime/api/runtime.proto [hunk 7]
```diff
@@ -11,6 +11,25 @@ option go_package = "github.com/rilldata/rill/runtime/api;api";
...
+      get: "/v1/instances/{instance_id}/tables/{table_name}/profile-columns"
+    };
+  }
+
+  rpc TableRows(RowsRequest) returns (RowsResponse) {
+    option (google.api.http) = {
+      get: "/v1/instances/{instance_id}/tables/{table_name}/rows"
+    };
+  }
+
```
Comment (begelundmuller) [inline runtime/api/runtime.proto:32]: Maybe move these below the explore APIs as discussed in our meeting

**Comment by nishantmonu51** [comment]
LGTM, can be merged once benjamin's comments are resolved.

File: runtime/api/runtime.proto (hunk 7 above)
Comment (egor-ryashin) [reply runtime/api/runtime.proto]: moved

File: runtime/server/table_api_test.go (hunk 6 above)
Comment (egor-ryashin) [reply runtime/server/table_api_test.go]: changed

File: runtime/api/runtime.proto (hunk 5 above)
Comment (egor-ryashin) [reply runtime/api/runtime.proto]: changed

**Review by begelundmuller — APPROVED** [review]

**Pull request merged**